metricbeat/module/system/network/network.go

```diff
@@ -104,46 +105,102 @@ func (m *MetricSet) Fetch(r mb.ReporterV2) error {
-		// convert network metrics from counters to gauges
+	if len(m.currentGaugeCounter) != 0 {
+
+		var totalNetworkInBytes, totalNetworkInPackets, totalNetworkOutBytes, totalNetworkOutPackets uint64
+		for _, iface := range m.currentGaugeCounter {
+			totalNetworkInBytes += iface.NetworkInBytes
+			totalNetworkInPackets += iface.NetworkInPackets
+			totalNetworkOutBytes += iface.NetworkOutBytes
+			totalNetworkOutPackets += iface.NetworkOutPackets
+		}
 		r.Event(mb.Event{
 			RootFields: mapstr.M{
 				"host": mapstr.M{
 					"network": mapstr.M{
 						"ingress": mapstr.M{
-							"bytes":   networkInBytes - m.prevCounters.prevNetworkInBytes,
-							"packets": networkInPackets - m.prevCounters.prevNetworkInPackets,
+							"bytes":   totalNetworkInBytes,
+							"packets": totalNetworkInPackets,
 						},
 						"egress": mapstr.M{
-							"bytes":   networkOutBytes - m.prevCounters.prevNetworkOutBytes,
-							"packets": networkOutPackets - m.prevCounters.prevNetworkOutPackets,
+							"bytes":   totalNetworkOutBytes,
+							"packets": totalNetworkOutPackets,
 						},
 					},
 				},
 			},
 		})
 	}
 
-	// update prevCounters
-	m.prevCounters.prevNetworkInBytes = networkInBytes
-	m.prevCounters.prevNetworkInPackets = networkInPackets
-	m.prevCounters.prevNetworkOutBytes = networkOutBytes
-	m.prevCounters.prevNetworkOutPackets = networkOutPackets
-
 	return nil
 }
 
+// Create a gauged difference between two numbers, taking into account rollover that might happen, and the current number might be lower.
+// The /proc/net/dev interface is defined in net/core/net-procfs.c,
+// where it prints the data from rtnl_link_stats64 defined in uapi/linux/if_link.h.
+// There's an extra bit of logic here: the underlying network device object in the kernel, net_device,
+// can define either ndo_get_stats64() or ndo_get_stats() as a metrics callback, with the latter returning an unsigned long (32 bit) set of metrics.
+// See dev_get_stats() in net/core/dev.c for context. The exact implementation depends upon the network driver.
+// For example, the tg3 network driver used by the broadcom network controller on my dev machine
+// uses 64 bit metrics, defined in the drivers/net/ethernet/broadcom/tg3.h,
+// with the ndo_get_stats64() callback defined in  net/ethernet/broadcom/tg3.c.
+// Long story short, we can't be completely sure if we're rolling over at max_u32 or max_u64.
+// if the previous value was > max_u32, do math assuming we've rolled over at max_u64.
+// On windows: This uses GetIfEntry: https://learn.microsoft.com/en-us/windows/win32/api/netioapi/ns-netioapi-mib_if_row2 which uses ulong64.
+// On Darwin we just call netstat.
+// I'm assuming rollover behavior is similar.
+func createGaugeWithRollover(current uint64, prev uint64) uint64 {
+	// base case: no rollover
+	if current > prev || current == prev {
+		return current - prev
+	}
+	debugf("Warning: Gauge has rolled over. Current value: %d, previous: %d", current, prev)
```

Comment:
```suggestion
```

if something goes wrong I think it would be useful to know which case we were in.